app/jobs/regular/process_post.rb

```diff
@@ -60,6 +60,26 @@ def extract_links(post)
       TopicLink.extract_from(post)
       QuotedPost.extract_from(post)
     end
+
+    def auto_tag(post)
+      topic_tags = post.topic.tags.pluck(:name)
+      changed = false
+
+      WordWatcher.words_for_action(:tag).each do |word, tags|
+        if SiteSetting.watched_words_regular_expressions? ? Regexp.new(word).match?(post.raw) : post.raw.include?(word)
+          topic_tags += tags.split(",")
+          changed = true
```

Comment:
I think we should only revise when there's an actual change in the list of tags. You could compare the initial list of tags to a list of tags you've built using the `WordWatcher`.